### PR TITLE
Implement ActivationMode= as an option to control networkd activation of the link

### PR DESCRIPTION
**Pull request by sil2100**
This is a continuation of PR: #9956 since Matt is no longer available to work on this. Please see the previous PR for history - I made the branch up-to-date with current master and (hopefully) addressed all the issues outlined in the previous reviews.

Please give me a sign if anything else needs tweaking. Thank you!

### CODE REVIEW

**Comment by yuwata** [comment]
Please do not include the merge commit in the PR.

File: man/systemd.network.xml [hunk 1]
```diff
@@ -225,6 +225,26 @@
...
+
+          <para>When set to <literal>on</literal>, networkd manages the device
+          normally and will bring it online as appropriate.</para>
+
+          <para>When set to <literal>manual</literal>, the device will be
+          configured but will remain in the UP/DOWN state it was when discovered,
+          and changing its state will need to be done by the administrator, with
+          a command such as <literal>ip link set up/down</literal>.</para>
+
+          <para>When <literal>no</literal>, networkd will attempt to keep the
```
Comment (yuwata) [inline man/systemd.network.xml]: no -> off

**Comment by sil2100** [comment]
Thank you for the review! I'm preparing some tests for the setting and will force push to this branch once I confirm the test cases are correct.

**Comment by sil2100** [comment]
I have now pushed all the changes as a single commit. Fixed the typo in the documentation and actually fixed a bug in the implementation, as the setting wasn't set properly due to an error in networkd-network-gperf.gperf.

I have added two test cases, hopefully those will be enough and feel correct - this is my first time working with the systemd test suite (and actually systemd itself). Thank you!

File: man/systemd.network.xml (hunk 1 above)
Comment (sil2100) [reply man/systemd.network.xml]: Fixed! Thank you!

**Comment by sil2100** [comment]
After a re-merge and push the tests seem to be now passing. What can I do to get this re-reviewed and considered? Thank you!

**Comment by xnox** [comment]
@yuwata would you be able to re-review this please?

File: man/systemd.network.xml [hunk 2]
```diff
@@ -225,6 +225,26 @@
...
       </varlistentry>
+      <varlistentry>
+        <term><varname>ActivationMode=</varname></term>
+        <listitem>
+          <para>Sets the activation mode (whether the device should be brought
+          up automatically or not) for the device. Accepts
+          <literal>on</literal>, <literal>off</literal>, or
+          <literal>manual</literal>. Defaults to <literal>on</literal>.</para>
+
+          <para>When set to <literal>on</literal>, networkd manages the device
```
Comment (poettering) [inline man/systemd.network.xml]: `networkd` → `<command>systemd-networkd</command>`

Comment (sil2100) [reply man/systemd.network.xml]: Changed!

**Comment by sil2100** [comment]
Resolved conflicts, resolved string issue, force pushed. Hopefully now it will be good to go!

**Comment by yuwata** [comment]
The files 25-activation-mode-off.network and 25-activation-mode-manual.network are not included in the commit. So, CentOS CIs are failed.

**Comment by sil2100** [comment]
Hopefully didn't forget any files this time!

File: test/test-network/systemd-networkd-tests.py [hunk 3]
```diff
@@ -1648,6 +1648,8 @@ class NetworkdNetworkTests(unittest.TestCase, Utilities):
         '23-active-slave.network',
         '24-keep-configuration-static.network',
         '24-search-domain.network',
+        '25-activation-mode-manual.network',
+        '25-activation-mode-off.network',
```
Comment (yuwata) [inline test/test-network/systemd-networkd-tests.py:1652]: I've added these lines.

**Comment by ddstreet** [comment]
Sorry that I'm quite late in jumping into this, but I'm a bit concerned about this PR...since there seems to have been a *long* history for this, spanning two previous PRs, maybe I'm totally misunderstanding what problem this is intending to solve... @sil2100 could you give a quick summary of what/why this is doing?

**Comment by xnox** [comment]
> Sorry that I'm quite late in jumping into this, but I'm a bit concerned about this PR...since there seems to have been a _long_ history for this, spanning two previous PRs, maybe I'm totally misunderstanding what problem this is intending to solve... @sil2100 could you give a quick summary of what/why this is doing?

It's about fixing the LP bug report https://bugs.launchpad.net/netplan/+bug/1664844

**Comment by ddstreet** [comment]
> It's about fixing the LP bug report https://bugs.launchpad.net/netplan/+bug/1664844

aaaaaaaah....it's about duplicating ifupdown's "manual" mode in networkd.

Ok I see the utility of having a "manual" mode for networkd.  However, what is the point of "on" or "off" modes?  To be clear, this PR will change the behavior of networkd's control of IFF_UP, with no backwards compatibility.  Meaning, currently, networkd will bring up all managed interfaces at start, but if an interface is taken IFF_UP down, networkd won't forcibly and repeatedly bring it back up.

With this PR, the default mode ("on") means administrators can no longer 'ip link set down' because networkd will immediately bring it back up, every time.  While that might be something that some people want, it does not seem like a good *default*.

Similarly, if this is set to "off", networkd will force the interface IFF_UP *down*, always, and nobody will ever be able to bring the interface up because networkd will take it back down again immediately.  I suppose maybe some people might want that?  But I'll be most people who might choose "off" *actually* want off at networkd start, but still want to be able to bring the interface up later manually.

Further, this commit isn't actually doing the advertised things at the right points.  It modifies link_carrier_gained() and link_carrier_lost(), but carrier != IFF_UP.  Someone can administratively bring up an interface without carrier.  And, carrier can be lost while IFF_UP is still up.  If a physical interface is IFF_UP and someone removes the physical cable, the carrier is going to drop, and it has nothing to do with IFF_UP.  Additionally, the way this is implemented, if IgnoreCarrierLoss=yes, this will *not* bring the interface back up, even if the reason for carrier loss was IFF_UP going down.

So, @sil2100, please do correct me if I'm still misunderstanding the goal, and the implementation, of this PR, but otherwise I think 1) the parameter needs to be rethought and 2) the implementation needs to be redone.

Personally, I would suggest a 5-way parameter:
ALWAYS_UP
START_UP
MANUAL
START_DOWN
ALWAYS_DOWN

where 'ALWAYS_UP' and 'ALWAYS_DOWN' mean that networkd will react to netlink IFF_UP flag changes and always move IFF_UP to the specified setting (again, NOT reacting to carrier changes).

then 'START_UP' and 'START_DOWN' would mean networkd brings IFF_UP up or down, respectively, when netowrkd starts - the default would be 'START_UP' which is networkd's current behavior.

finally 'MANUAL' means networkd never changes IFF_UP.  It's up to something/someone else to change the interface's IFF_UP.

Also, "ActivationMode" is IMHO a *highly* misleading parameter name, since it doesn't actually refer to only networkd's initial startup "activation".  It would be better to call it "AdministrativeState" since IFF_UP is literally referred to as "Administrative state":
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/networking/operstates.rst#n12

**Comment by poettering** [comment]
I think @ddstreet's points make sense. Given that this requires some rework I think we should drop this from the milestone, to not delay that needlessly. Still would love to merge this before the next release, but for that it would need to be updated soon.

**Comment by sil2100** [comment]
Thank you for your reviews Dan, Lennart! I think your arguments make sense. This here was basically a very old PR, created by cyphermox a while ago, with me only recently picking it up and re-submitting after applying requested changes.
Let me try getting to this this week.

**Comment by poettering** [comment]
Also needs rebase.

Let's drop this from the milestone. I see no reason to delay releases for this. We can merge this whenever it's ready.

**Comment by keszybz** [comment]
#16228 is very similar but much more complete at this point. @sil2100 could you take a look at #16228 and in particular check if it solves your usecases.

**Comment by sil2100** [comment]
Thank you for pointing out the new PR! First of all, apologies for not attending this in a long time - have been preempted by other tasks. Quickly looking at #16228 it does seem like the solution to our initial problem, just solved in a more flexible way (per Dan's earlier proposed design). I'll take a look at the diff a bit more in a moment but I think I'll close this PR in favor of the other.

**Comment by sil2100** [comment]
Let me close this one off in favor of the other PR, we can continue there.

**Pull request closed without merging**